File: admin-client/src/lib/services/early-access.ts
import { API_URL } from '@/lib/config';
import { getAuthToken } from './auth';

/**
 * Represents an early access registration entry
 */
export interface EarlyAccess {
  id: number;
  email: string;
  referrer: string;
  created_at: string;
  updated_at: string;
}

/**
 * Interface for the response from the early access API
 */
export interface GetEarlyAccessResponse {
  registrations: EarlyAccess[];
}

/**
 * Creates headers with authentication for API requests
 */
const headers = () => ({
  'Authorization': `Bearer ${getAuthToken()}`,
  'Content-Type': 'application/json',
});

/**
 * Fetches all early access registrations from the API
 * @returns Promise containing the list of early access registrations
 */
export const getEarlyAccessRegistrations = async (): Promise<EarlyAccess[]> => {
  const response = await fetch(`${API_URL}/admin/early-access`, {
    headers: headers(),
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(error || 'Failed to fetch early access registrations');
  }

  const data = await response.json();
  
  // Return the registrations with consistent structure
  return data || [];
}; 

File: server/models/early_access.go
package models

import (
	"time"
)

// EarlyAccess represents a user who has signed up for early access
type EarlyAccess struct {
	ID        int       `json:"id"`
	Email     string    `json:"email"`
	Referrer  string    `json:"referrer,omitempty"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// EarlyAccessRequest represents the data sent from the frontend
type EarlyAccessRequest struct {
	Email    string `json:"email"`
	Referrer string `json:"referrer,omitempty"`
}
